sql/core/src/test/scala/org/apache/spark/sql/streaming/DeduplicationSuite.scala

```diff
@@ -0,0 +1,235 @@
+      CheckLastBatch((10 -> 5)), // 5 items (10 to 14) after deduplication
+      // states in aggregation in [15, 20) and [25, 30) (2 windows, note aggregation uses the end of
+      // window to evict items, so [15, 20) is still in the state store)
+      // states in deduplication is 25
+      assertNumStateRows(total = Seq(2L, 1L), updated = Seq(0L, 0L)),
+
+      AddData(inputData, 10), // Should not emit anything as data less than watermark
+      CheckLastBatch(),
+      assertNumStateRows(total = Seq(2L, 1L), updated = Seq(0L, 0L)),
+
+      AddData(inputData, 40), // Advance watermark to 30 seconds
+      CheckLastBatch(),
+      // states in aggregation in [15, 20), [25, 30) and [40, 45)
+      // states in deduplication is 25 and 40,
+      assertNumStateRows(total = Seq(3L, 2L), updated = Seq(1L, 1L)),
+
+      AddData(inputData, 40), // Emit items less than watermark and drop their state
+      CheckLastBatch((15 -> 1), (25 -> 1)),
+        // states in aggregation in [40, 45)
+      // states in deduplication is 40,
+      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(0L, 0L))
+    )
+  }
+
+  test("deduplication with aggregation - update") {
+    val inputData = MemoryStream[(String, Int)]
+    val result = inputData.toDS()
+      .dropDuplicates()
+      .groupBy($"_1")
+      .agg(sum($"_2"))
+      .as[(String, Long)]
+
+    testStream(result, Update)(
+      AddData(inputData, "a" -> 1),
+      CheckLastBatch("a" -> 1L),
+      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(1L, 1L)),
+      AddData(inputData, "a" -> 1), // Dropped
+      CheckLastBatch(),
+      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(0L, 0L)),
+      AddData(inputData, "a" -> 2),
+      CheckLastBatch("a" -> 3L),
+      assertNumStateRows(total = Seq(1L, 2L), updated = Seq(1L, 1L)),
+      AddData(inputData, "b" -> 1),
+      CheckLastBatch("b" -> 1L),
+      assertNumStateRows(total = Seq(2L, 3L), updated = Seq(1L, 1L))
+    )
+  }
+
+  test("deduplication with aggregation - complete") {
+    val inputData = MemoryStream[(String, Int)]
+    val result = inputData.toDS()
+      .dropDuplicates()
+      .groupBy($"_1")
+      .agg(sum($"_2"))
+      .as[(String, Long)]
+
+    testStream(result, Complete)(
+      AddData(inputData, "a" -> 1),
+      CheckLastBatch("a" -> 1L),
+      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(1L, 1L)),
+      AddData(inputData, "a" -> 1), // Dropped
+      CheckLastBatch("a" -> 1L),
+      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(0L, 0L)),
+      AddData(inputData, "a" -> 2),
+      CheckLastBatch("a" -> 3L),
+      assertNumStateRows(total = Seq(1L, 2L), updated = Seq(1L, 1L)),
+      AddData(inputData, "b" -> 1),
+      CheckLastBatch("a" -> 3L, "b" -> 1L),
+      assertNumStateRows(total = Seq(2L, 3L), updated = Seq(1L, 1L))
+    )
+  }
+
+  private def assertNumStateRows(total: Seq[Long], updated: Seq[Long]): AssertOnQuery =
```

Comment:
These methods seems to be common across StreamingAggregationSuite, MapGroupsWithStateSuite and this one. Can you make a trait?